consensus/state_processing/src/per_epoch_processing/base/rewards_and_penalties.rs

```diff
@@ -78,17 +80,47 @@ pub fn get_attestation_deltas<T: EthSpec>(
     validator_statuses: &ValidatorStatuses,
     spec: &ChainSpec,
 ) -> Result<Vec<AttestationDelta>, Error> {
+    get_attestation_deltas_subset(state, validator_statuses, None, spec)
+        .map(|deltas_map| deltas_map.into_iter().map(|(_, deltas)| deltas).collect())
+}
+
+/// Apply rewards for participation in attestations during the previous epoch.
+/// If `maybe_validators_subset` specified, only the deltas for the specified validator subset is
+/// returned, otherwise deltas for all validators are returned.
+///
+/// Returns a vec of validator indices to `AttestationDelta`.
+pub fn get_attestation_deltas_subset<T: EthSpec>(
+    state: &BeaconState<T>,
+    validator_statuses: &ValidatorStatuses,
+    maybe_validators_subset: Option<&Vec<usize>>,
+    spec: &ChainSpec,
+) -> Result<Vec<(usize, AttestationDelta)>, Error> {
     let previous_epoch = state.previous_epoch();
     let finality_delay = state
         .previous_epoch()
         .safe_sub(state.finalized_checkpoint().epoch)?
         .as_u64();
 
-    let mut deltas = vec![AttestationDelta::default(); state.validators().len()];
+    let mut deltas = HashMap::new();
```

Comment:
HashMaps are quite a bit slower than Vecs for indexing, and the _sort_ at the end is also `O(n log n)`. I think we're sacrificing a bit too much performance here in the all-validators case to save time in the subset-of-validators case.

Instead, I'd like to propose that we keep `get_attestation_deltas` the same as it was previously, and take the performance hit when computing rewards for a small number of validators. On mainnet the max number of phase0 validators is 250k in epoch [74239](https://beaconcha.in/epoch/74239) right before the Altair fork. Computing rewards for these 250k validators probably only takes ~50-250ms, which is not a huge saving in absolute terms, and will usually be smaller than the time required for state reconstruction (500ms+ at least). If we slow down the subset-validators case the cost (call it `subset_slowdown`) would be paid only by users querying small numbers of validators (`slowdown x number_of_user_queries`).

By comparison, if we slow down epoch processing in general (the all-validators case), then this slowdown (`all_slowdown`) will be paid by every archive node, at least once for every epoch transition (`number_of_archive_nodes x 74k epochs x all_slowdown`). The main consumers of this API (block explorers) are also likely to be making all-validators queries impacted by this slowdown. Thereofore I think the right balance is to prioritise the all-validators case for now, and we can come back to speed-up individual validator requests later if we desire (e.g. if the GUI starts using these endpoints).

PS: if we do want to keep the map approach, we should use a `BTreeMap` so that it retains its order during the calculation and we can avoid needing to sort it at the end.